lib/ddtrace/configuration.rb

```diff
@@ -9,6 +10,18 @@ module Datadog
   module Configuration
     extend Forwardable
 
+    # Used to ensure that @components initialization/reconfiguration is performed one-at-a-time, by a single thread.
+    #
+    # This is important because components can end up being accessed from multiple application threads (for instance on
+    # a threaded webserver), and we don't want their initialization to clash (for instance, starting two profilers...).
+    #
+    # Note that a Mutex **IS NOT** reentrant: the same thread cannot grab the same Mutex more than once.
+    # This means below we are careful not to nest calls to methods that grab the lock.
```

Comment:
Currently our trace does perform reentrant calls to component initialization.
One example is calling `Datadog.logger` when any part of the component initialization reports an error message.

If we want to introduce this change today, I don't think we can have this mutex as non-reentrant: the it is a requirement for `ddtrace` to not prevent the host application from continuing its work, even when `ddtrace` fails.

Reply:
Yeah, I'll admit I hesitated a bit between going for `Mutex` or `Monitor`.

I believe the `logger` example you quoted cannot happen -- the `logger`  is built to never trigger the initialization of the components (`if current_components` will never be `true` until the components are properly and fully initialized).

> If we want to introduce this change today, I don't think we can have this mutex as non-reentrant: the it is a requirement for ddtrace to not prevent the host application from continuing its work, even when ddtrace fails.

If the use-case is "it should never break customer code", I was considering an alternative -- what if we caught the resulting `ThreadError` and just logged an error message instead?

I'm somewhat concerned that going with the [thread-safe `Monitor`](https://rubyapi.org/3.0/o/monitor) instead may just silently hide issues, whereas if we handled the exception and printed an error we could identify issues a lot easier.

Reply:
We can try catching `ThreadError`, that might do it. 👍

Reply:
Added handling of `ThreadError` in fdc0d5848b6af87c754161b61d7ad79795e70fbe, let me know your thoughts!

Reply:
In particular, this is the second time (the other being https://github.com/DataDog/dd-trace-rb/pull/1310) that I've added a user-visible "dear user, oops we messed up! do you mind telling us about it so we can fix it?" message.

I think being clear/open about these allows us to better spot and improve the codebase, but this is a somewhat new pattern in dd-trace-rb so I'm curious what the rest of the team thinks about it.